File: Arrays/contains-duplicate-ii/contains_duplicate_ii.py
#  Project : SDE Placement Questions
#  Filename : contains_duplicate_ii.py
#  Author : thameem
#  Current modification time : Tue, 11 Oct 2022 at 11:07 pm India Standard Time
#  Last modified time : Tue, 11 Oct 2022 at 11:07 pm India Standard Time

from typing import List


def contains_nearby_duplicate(nums: List[int], k: int) -> bool:
    duplicate_values = {}

    for index, num in enumerate(nums):
        if num in duplicate_values and abs(index - duplicate_values[num]) <= k:
            return True
        else:
            duplicate_values[num] = index

    return False


if __name__ == '__main__':
    print(contains_nearby_duplicate(nums=[1, 2, 3, 1], k=3))

File: Arrays/contains-duplicate-ii/contains_duplicate_ii.cpp
#include <iostream>
#include <vector>
#include <map>
using namespace std;

bool contains_nearby_duplicate(vector<int> &nums, int k)
{
    map<int, int> duplicate_values;

    for (int i = 0; i < nums.size(); i++)
    {
        if (duplicate_values.count(nums[i]) != 0 and abs(i - duplicate_values[nums[i]]) <= k)
        {
            return true;
        }
        else
        {
            duplicate_values[nums[i]] = i;
        }
    }

    return false;
}

int main()
{
    vector<int> nums = {1, 2, 3, 1};
    int k = 3;
    cout << contains_nearby_duplicate(nums, k);
    return 0;
}